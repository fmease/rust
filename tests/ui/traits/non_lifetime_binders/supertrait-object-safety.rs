// compile-flags: -Ztrait-solver=next

#![feature(non_lifetime_binders)]
//~^ WARN the feature `non_lifetime_binders` is incomplete

trait Foo: for<T> Bar<T> {}

trait Bar<T: ?Sized> {
    fn method(&self) {}
}

fn needs_bar(x: &(impl Bar<i32> + ?Sized)) {
    x.method();
}

impl Foo for () {}

impl<T: ?Sized> Bar<T> for () {}

fn main() {
    let x: &dyn Foo = &();
    //~^ ERROR the type `&dyn Foo` is not well-formed
    //~| ERROR mismatched types
    needs_bar(x);
    //~^ ERROR the type `&dyn Foo` is not well-formed
    //~| ERROR the type `dyn Foo` is not well-formed
    //~| ERROR the trait bound `dyn Foo: Bar<i32>` is not satisfied
}
