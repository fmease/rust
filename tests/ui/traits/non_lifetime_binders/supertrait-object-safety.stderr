warning: the feature `non_lifetime_binders` is incomplete and may not be safe to use and/or cause compiler crashes
  --> $DIR/supertrait-object-safety.rs:3:12
   |
LL | #![feature(non_lifetime_binders)]
   |            ^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #108185 <https://github.com/rust-lang/rust/issues/108185> for more information
   = note: `#[warn(incomplete_features)]` on by default

error[E0308]: mismatched types
  --> $DIR/supertrait-object-safety.rs:21:23
   |
LL |     let x: &dyn Foo = &();
   |            --------   ^^^ expected `&dyn Foo`, found `&()`
   |            |
   |            expected due to this
   |
   = note: expected reference `&dyn Foo`
              found reference `&()`

error: the type `&dyn Foo` is not well-formed
  --> $DIR/supertrait-object-safety.rs:21:12
   |
LL |     let x: &dyn Foo = &();
   |            ^^^^^^^^

error[E0277]: the trait bound `dyn Foo: Bar<i32>` is not satisfied
  --> $DIR/supertrait-object-safety.rs:24:15
   |
LL |     needs_bar(x);
   |     --------- ^ the trait `Bar<i32>` is not implemented for `dyn Foo`
   |     |
   |     required by a bound introduced by this call
   |
   = help: the trait `Bar<T>` is implemented for `()`
note: required by a bound in `needs_bar`
  --> $DIR/supertrait-object-safety.rs:12:24
   |
LL | fn needs_bar(x: &(impl Bar<i32> + ?Sized)) {
   |                        ^^^^^^^^ required by this bound in `needs_bar`

error: the type `dyn Foo` is not well-formed
  --> $DIR/supertrait-object-safety.rs:24:5
   |
LL |     needs_bar(x);
   |     ^^^^^^^^^

error: the type `&dyn Foo` is not well-formed
  --> $DIR/supertrait-object-safety.rs:24:15
   |
LL |     needs_bar(x);
   |               ^

error: aborting due to 5 previous errors; 1 warning emitted

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
